pkg/doc/did/doc.go

```diff
@@ -714,49 +801,48 @@ func (doc *Doc) JSONBytes() ([]byte, error) {
 		context = doc.Context[0]
 	}
 
-	publicKeys, err := populateRawPublicKeys(context, doc.PublicKey)
+	publicKeys, err := populateRawPublicKeys(context, doc.processingMeta.base, doc.PublicKey)
 	if err != nil {
 		return nil, fmt.Errorf("JSON unmarshalling of Public Key failed: %w", err)
 	}
 
-	auths, err := populateRawVerificationMethods(context, doc.ID, doc.Authentication)
+	auths, err := populateRawVerificationMethods(context, doc.processingMeta.base, doc.ID, doc.Authentication)
 	if err != nil {
 		return nil, fmt.Errorf("JSON unmarshalling of Authentication failed: %w", err)
 	}
 
-	assertionMethods, err := populateRawVerificationMethods(context, doc.ID, doc.AssertionMethod)
+	assertionMethods, err := populateRawVerificationMethods(context, doc.processingMeta.base, doc.ID, doc.AssertionMethod)
 	if err != nil {
 		return nil, fmt.Errorf("JSON unmarshalling of AssertionMethod failed: %w", err)
 	}
 
-	capabilityDelegations, err := populateRawVerificationMethods(context, doc.ID, doc.CapabilityDelegation)
+	capabilityDelegations, err := populateRawVerificationMethods(context, doc.processingMeta.base, doc.ID,
+		doc.CapabilityDelegation)
 	if err != nil {
 		return nil, fmt.Errorf("JSON unmarshalling of CapabilityDelegation failed: %w", err)
 	}
 
-	capabilityInvocations, err := populateRawVerificationMethods(context, doc.ID, doc.CapabilityInvocation)
+	capabilityInvocations, err := populateRawVerificationMethods(context, doc.processingMeta.base, doc.ID,
+		doc.CapabilityInvocation)
 	if err != nil {
 		return nil, fmt.Errorf("JSON unmarshalling of CapabilityInvocation failed: %w", err)
 	}
 
-	keyAgreements, err := populateRawVerificationMethods(context, doc.ID, doc.KeyAgreement)
+	keyAgreements, err := populateRawVerificationMethods(context, doc.processingMeta.base, doc.ID, doc.KeyAgreement)
 	if err != nil {
 		return nil, fmt.Errorf("JSON unmarshalling of KeyAgreement failed: %w", err)
 	}
 
 	raw := &rawDoc{
-		Context:              doc.Context,
-		ID:                   doc.ID,
-		PublicKey:            publicKeys,
-		Authentication:       auths,
-		AssertionMethod:      assertionMethods,
-		CapabilityDelegation: capabilityDelegations,
-		CapabilityInvocation: capabilityInvocations,
-		KeyAgreement:         keyAgreements,
-		Service:              populateRawServices(doc.Service),
-		Created:              doc.Created,
-		Proof:                populateRawProofs(context, doc.Proof),
-		Updated:              doc.Updated,
+		Context: doc.Context, ID: doc.ID, PublicKey: publicKeys,
+		Authentication: auths, AssertionMethod: assertionMethods, CapabilityDelegation: capabilityDelegations,
+		CapabilityInvocation: capabilityInvocations, KeyAgreement: keyAgreements,
+		Service: populateRawServices(doc.Service, doc.processingMeta.base), Created: doc.Created,
+		Proof: populateRawProofs(context, doc.processingMeta.base, doc.Proof), Updated: doc.Updated,
+	}
+
+	if doc.processingMeta.base != "" {
```

Comment:
@fqutishat DID Core specifies the default transformation behavior for relative URIs into absolute URIs. I would think we would follow this behavior when base isn't present.

See https://www.w3.org/TR/did-core/#relative-did-urls

Reply:
In essence, I would think baseURI isn't empty but rather populated either due to JSON-LD base property or defaulting to the DID URI.

Reply:
I didn't change the behaviour when base not exist it's the same

Reply:
@troyronda @fqutishat 

> defaulting to the DID URI

[Relative URL dereferencing](https://www.w3.org/TR/did-core/#relative-did-urls) is defined over the DID document model, not over a core representation. After you parse the representation, whether it is in JSON-LD or just JSON, you end up with the DID document model wherein some of the references may still be relative. It is here we must take the base URI as the DID subject.

We may choose to opt out of having relative references after parsing the did document into the abstract model for [reasons] but I just wanted to make that distinction. Hopefully it helps.

Reply:
done

Reply:
@llorllale

Reply:
@fqutishat is there a test that shows how the document is serialized to bytes?

Reply:
Regardless, I consider the followup #2281 as the story to track to refactor all this.

Reply:
FYI just tested by adding `validDocWithBase` to `TestJSONConversion`. ParseDocument and JSONBytes are symmetrical and the serialization adds `@base`

Reply:
Done

Reply:
>and the serialization adds `@base`

@llorllale @fqutishat I assume that we aren't adding the base property when serializing peer DID documents right?

Reply:
This PR is adding it.

Reply:
@fqutishat @llorllale  We shouldn't be implicitly adding base property. This property isn't a default property for DID methods.

Reply:
I'm not adding that

Reply:
if @base is exist in DID document I will add it

Reply:
if not exist before parsing I will not add it

Reply:
this test will confirm TestJSONConversion

Reply:
https://github.com/hyperledger/aries-framework-go/compare/87947039894ac90b9da3ef3dbc8ca4f2f21020cf..e86b70555480fba40df163f17f38ccc6d80a0525

Reply:
@troyronda 

> @fqutishat @llorllale We shouldn't be implicitly adding base property when we output DID documents. This property isn't a default property for DID methods.

Yes. Followups will be done to rectify this as part of #2281